pkg/gadgets/trace/byob/tracer/tracer.go

```diff
@@ -0,0 +1,328 @@
+	}
+	if err != nil {
+		return fmt.Errorf("create BPF map reader: %w", err)
+	}
+
+	// Attach programs
+	for progName, p := range t.spec.Programs {
+		if p.Type == ebpf.Kprobe && strings.HasPrefix(p.SectionName, "kprobe/") {
+			l, err := link.Kprobe(p.AttachTo, t.collection.Programs[progName], nil)
+			if err != nil {
+				return fmt.Errorf("attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		} else if p.Type == ebpf.Kprobe && strings.HasPrefix(p.SectionName, "kretprobe/") {
+			l, err := link.Kretprobe(p.AttachTo, t.collection.Programs[progName], nil)
+			if err != nil {
+				return fmt.Errorf("attach BPF program %q: %w", progName, err)
+			}
+			t.links = append(t.links, l)
+		}
+	}
+
+	return nil
+}
+
+func (t *Tracer) run(gadgetCtx gadgets.GadgetContext) {
+	d := t.decoderFactory()
+	for {
+		var rawSample []byte
+
+		if t.ringbufReader != nil {
+			record, err := t.ringbufReader.Read()
+			if err != nil {
+				if errors.Is(err, ringbuf.ErrClosed) {
+					// nothing to do, we're done
+					return
+				}
+				gadgetCtx.Logger().Errorf("read ring buffer: %w", err)
+				return
+			}
+			rawSample = record.RawSample
+		} else if t.perfReader != nil {
+			record, err := t.perfReader.Read()
+			if err != nil {
+				if errors.Is(err, perf.ErrClosed) {
+					return
+				}
+				gadgetCtx.Logger().Errorf("read perf ring buffer: %w", err)
+				return
+			}
+
+			if record.LostSamples != 0 {
+				gadgetCtx.Logger().Warnf("lost %d samples", record.LostSamples)
+				continue
+			}
+			rawSample = record.RawSample
+		} else {
+			gadgetCtx.Logger().Error("neither ring buffer or perf ring buffer was found")
+			return
+		}
+
+		if uint32(len(rawSample)) < t.mapSizes[t.printMap] {
+			gadgetCtx.Logger().Errorf("read ring buffer: len(RawSample)=%d!=%d",
+				len(rawSample),
+				t.mapSizes[t.printMap])
+			return
+		}
+
+		// FIXME: DecodeBtfBinary has a bug with non-NULL-terminated strings.
+		// For now, ensure the problem does not happen in ebpf
+
+		result, err := d.DecodeBtfBinary(gadgetCtx.Context(), t.valueStruct, rawSample[:t.mapSizes[t.printMap]])
+		if err != nil {
+			gadgetCtx.Logger().Errorf("decoding btf: %w", err)
+			return
+		}
+		b, err := json.Marshal(result)
+		if err != nil {
+			gadgetCtx.Logger().Errorf("encoding json: %w", err)
+			return
+		}
+
+		event := types.Event{
+			Event: eventtypes.Event{
+				Type: eventtypes.NORMAL,
+			},
+			WithMountNsID: eventtypes.WithMountNsID{MountNsID: 0},
+			Payload:       fmt.Sprintf("%+v", string(b)),
+		}
+
+		if mnt_ns_id_str, ok := result["mnt_ns_id"]; ok {
+			if mnt_ns_id, ok := mnt_ns_id_str.(uint64); ok {
+				event.MountNsID = mnt_ns_id
+			}
+		}
+		t.eventCallback(&event)
+	}
+}
+
+func (t *Tracer) Run(gadgetCtx gadgets.GadgetContext) error {
+	go t.run(gadgetCtx)
+	gadgetcontext.WaitForTimeoutOrDone(gadgetCtx)
+
+	return nil
+}
+
+func (t *Tracer) SetMountNsMap(mountnsMap *ebpf.Map) {
+	t.config.MountnsMap = mountnsMap
+}
```

Comment:
For future discussion: I think we need to make this more dynamic, not all gadgets will allow this. One idea is to change 

https://github.com/inspektor-gadget/inspektor-gadget/blob/829717b6a68d664ba3fc9c919638decc2495fd9f/pkg/operators/kubemanager/kubemanager.go#L42-L44

to have a second `CanSetMountNsMap() bool` method to check if the given gadget actually accepts a mount ns map.

/cc @flyth